### PR TITLE
Implement Rails specific Clowder initialization

### PR DESCRIPTION
**Pull request by RoamingNoMaD**
This PR contains the implementation of the `RailsConfig` class and an engine, which loads the necessary configuration of a Rails app upon initialization. 

These methods were extracted from the `compliance-backend` repository in the following [PR](https://github.com/RedHatInsights/compliance-backend/pull/1664 ).

### CODE REVIEW

File: lib/clowder-common-ruby.rb [hunk 1]
```diff
@@ -4,3 +4,7 @@
 require 'clowder-common-ruby/version'
 require 'clowder-common-ruby/config'
 require 'clowder-common-ruby/types'
+require 'clowder-common-ruby/engine'
```
Comment (skateman) [inline lib/clowder-common-ruby.rb:7]: This should not be the part of the base gem, only rails apps should load it.

File: lib/clowder-common-ruby/config.rb [hunk 2]
```diff
@@ -84,4 +84,158 @@ def private_dependency_endpoints
       end
     end
   end
+
+  class RailsConfig
```
Comment (skateman) [inline lib/clowder-common-ruby/config.rb:88]: This class should be in a separate file.

**Review by skateman — CHANGES_REQUESTED** [review]

File: lib/clowder-common-ruby/rails_config.rb [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        rbac: rbac_config,
+        redis: redis_config,
+        clowder_config_enabled: true,
+        prometheus_exporter_port: first_config&.metricsPort,
+        tlsCAPath: first_config.tlsCAPath
+      }
+
+      if first_config.dependency_endpoints.present?
+        first_config.dependency_endpoints.each do |name, endpoint|
+          key = "#{name}".underscore.to_sym
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:160]:
```suggestion
          key = name.underscore.to_sym
```

File: lib/clowder-common-ruby/rails_config.rb [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+      kafka_server_config = rails_config.configure_kafka
+
+      # Cloudwatch
+      cloudwatch_config = rails_config.configure_cloudwatch
+
+      # RBAC
+      rbac_config = rails_config.configure_rbac
+
+      # Redis
+      redis_config = rails_config.configure_redis
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:146]: I am not sure about these allocations, you could definitely move these method calls to the right side of the hash values below. Also regarding the `rails_config`, what about not having it at all? Just move all methods as singleton methods, taking one argument (the first config) that they actually need.

```
  first_config = ClowderCommonRuby::Config.load
  clowder_config = {
    kafka: configure_kafka(first_config),
    ...
```

File: lib/clowder-common-ruby/rails_config.rb [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+      clowder_config = {
+        kafka: kafka_server_config,
+        logging: cloudwatch_config,
+        rbac: rbac_config,
+        redis: redis_config,
+        clowder_config_enabled: true,
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:153]: I guess this can be omitted :thinking:

File: lib/clowder-common-ruby/rails_config.rb [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        prometheus_exporter_port: first_config&.metricsPort,
+        tlsCAPath: first_config.tlsCAPath
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:155]: let's stick with snake case everywhere

File: lib/clowder-common-ruby/rails_config.rb [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+      }
+
+      if first_config.dependency_endpoints.present?
+        first_config.dependency_endpoints.each do |name, endpoint|
+          key = "#{name}".underscore.to_sym
+          clowder_config[key] = rails_config.configure_endpoint(endpoint.dig('service'))
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:161]: I would not put them directly into the config, but under `:endpoints`

File: lib/clowder-common-ruby/rails_config.rb [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        end
+      end
+      
+      if first_config.private_dependency_endpoints.present?
+        first_config.private_dependency_endpoints.each do |name, endpoint|
+          key = "#{name}".underscore.to_sym
+          clowder_config[key] = rails_config.configure_endpoint(endpoint.dig('service'))
+        end
+      end
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:170]: In general these two steps can be done during the hash creation above using `each_with_object` and you don't really need to adjust the hash afterwards:

```
def self.build_endpoints(endpoints)
  endpoints.&each_with_object({}) do |name, endpoint, obj| # not sure about the name-endpoint destruct here, maybe (name, endpoint), obj ... try it
    key = ...
    obj[key] = configure_endpoint(endpoint.dig('service'))
  end
end

def self.to_h
  clowder_config = {
     ...
     endpoints: build_endpoints(first_config&.dependency_endpoints),
     private_endpoints: build_endpoints(first_config&.private_dependency_endpoints)
     ...
  }
end
```

**Review by skateman — CHANGES_REQUESTED** [review]

File: lib/clowder-common-ruby.rb [hunk 9]
```diff
@@ -4,3 +4,6 @@
 require 'clowder-common-ruby/version'
 require 'clowder-common-ruby/config'
 require 'clowder-common-ruby/types'
+
+module ClowderCommonRuby
+end
```
Comment (skateman) [inline lib/clowder-common-ruby.rb:9]: I don't think you really need these...

**Review by skateman — CHANGES_REQUESTED** [review]

File: lib/clowder-common-ruby/rails_config.rb [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    def initialize
+      @first_config = ClowderCommonRuby::Config.load
+    end
+    
+    # Map for Kafka configuration
+    def self.configure_kafka(first_config)
+      kafka_config = kafka_security(first_config)
+
+      kafka_config[:brokers] = kafka_brokers(first_config) || ''
+      kafka_config[:kafka_topics] = configure_kafka_topics(first_config)
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:14]:
```suggestion
      kafka_config[:kafka_topics] = kafka_topics(first_config)
```

File: lib/clowder-common-ruby/rails_config.rb [hunk 11]
```diff
@@ -0,0 +1,136 @@
+module ClowderCommonRuby
+  class RailsConfig
+    attr_reader :first_config
+    
+    def initialize
+      @first_config = ClowderCommonRuby::Config.load
+    end
+
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:8]: No longer needed

**Review by skateman — CHANGES_REQUESTED** [review]
Please drop the `first_config` word and just use `config` instead ... makes more sense.

File: config/initializers/test.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+require 'clowder-common-ruby/rails_config'
+
+if ClowderCommonRuby::Config.clowder_enabled?
+  config = ClowderCommonRuby::RailsConfig.to_hash
+
+  Settings.add_source!(config)
+  Settings.reload!
+end
```
Comment (skateman) [inline config/initializers/test.rb:8]: The name of this file should be something like `0-clowder-rails.rb`.

File: lib/clowder-common-ruby/rails_config.rb [hunk 13]
```diff
@@ -0,0 +1,123 @@
+module ClowderCommonRuby
+  class RailsConfig
+    attr_reader :config
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:3]: No longer used, no longer needed.

File: lib/clowder-common-ruby/rails_config.rb [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    
+    # Map for Kafka configuration
+    def self.configure_kafka(config)
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:6]: When defining singleton methods, you can simply put them into a `class << self` block:
```ruby
class BlaBla
  class << self
    def to_h
       ...
    end

    # Strong feature of this syntax is the singleton-private methods.
    private

    def configure_kafka(config)
      ...
    end
  end
```

File: lib/clowder-common-ruby/rails_config.rb [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+      build_kafka_security(config).merge(
+        brokers: build_kafka_brokers(config) || '',
+        topics: build_kafka_topics(config) 
+      )
+    end
+
+    # Map for Kafka security settings
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:13]: Kafka security settings hash

File: lib/clowder-common-ruby/rails_config.rb [hunk 16]
```diff
@@ -0,0 +1,123 @@
+module ClowderCommonRuby
+  class RailsConfig
+    attr_reader :config
+    
+    # Map for Kafka configuration
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:5]: Kafka configuration hash

File: lib/clowder-common-ruby/rails_config.rb [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    def self.configure_kafka(config)
+      build_kafka_security(config).merge(
+        brokers: build_kafka_brokers(config) || '',
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:8]: I would like to see a comment that explains why the `|| ''` is required.

File: lib/clowder-common-ruby/rails_config.rb [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+        sasl_username: server_config&.dig('sasl', 'username'),
+        sasl_password: server_config&.dig('sasl', 'password'),
+        sasl_mechanism: server_config&.dig('sasl', 'saslMechanism'),
+        security_protocol: server_config&.dig('sasl', 'securityProtocol'),
+        ssl_ca_location: build_kafka_certificate(cacert)
+      }.compact
+    end
+
+    def self.build_kafka_certificate(cacert)
+      return unless cacert.present?
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:35]: Explain in a comment what this does.

File: lib/clowder-common-ruby/rails_config.rb [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+
+      {
+        host: "#{endpoint.hostname}:#{port}",
+        scheme: scheme,
+        url: "#{scheme}://#{host}"
+      }
+    end
+
+    # Nested map for Rails configuration
+    def self.to_hash
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:109]:
```suggestion
    def self.to_h
```

File: lib/clowder-common-ruby/rails_config.rb [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+
+    # Map for configurations of all endpoints
+    def self.configure_endpoints(config, field)
+      config.try(field)&.each_with_object({}) do |(name, endpoint), obj|
+        key = name.underscore.to_sym
+        obj[key] = build_endpoint(config, endpoint.dig('service'))
+      end
+    end
+
+    # Map for configuration of a single endpoint
```
Comment (skateman) [inline lib/clowder-common-ruby/rails_config.rb:96]: Endpoint configuration hash

**Review by skateman — CHANGES_REQUESTED** [review]
1. map -> hash 
2. `class << self` syntax
3. `to_hash` -> `to_h`

**Comment by skateman** [comment]
Please add a description to the PR and link the counterpart, same for the counterpart...

**Review by vkrizan — APPROVED** [review]
Looks great Jany!

File: config/initializers/0_clowder_rails.rb [hunk 21]
```diff
@@ -0,0 +1,17 @@
+require 'clowder-common-ruby/rails_config'
+
+if ClowderCommonRuby::Config.clowder_enabled?
```
Comment (skateman) [inline config/initializers/0_clowder_rails.rb:3]: Let's make sure that `Settings` is available for the current context:
```suggestion
if ClowderCommonRuby::Config.clowder_enabled? && defined?(Settings)
```
cc @vkrizan so you sleep better :wink:

File: lib/clowder-common-ruby/version.rb [hunk 22]
```diff
@@ -1,3 +1,3 @@
 module ClowderCommonRuby
-  VERSION = '0.2.6'.freeze
+  VERSION = '0.3.0'.freeze
```
Comment (skateman) [inline lib/clowder-common-ruby/version.rb:2]: You'll need a rebase here.

**Review by skateman — APPROVED** [review]
Great job! Just please adjust the remaining two things and this can be merged

**Pull request merged**